### PR TITLE
Added support for converting from JPG to TFREC and augmenting 10% of images

### PR DESCRIPTION
**Pull request by GeorgeFAeillo**
- This commit includes a function that augments 10% of jpg files in a file path by rotating them and mirror flipping them randomly
- It also includes helper functions taken from Kaggle that should support the use of TFREC to JPEG and helper functions relating to bytes and string
-- Creating this pull request as a way for central documentation relating to these changes, these should not be approved without further testing and discussion

### CODE REVIEW

File: Code/monet_cyclegan/utils.py [hunk 1]
```diff
@@ -284,4 +285,66 @@ def log_args(args: Namespace) -> None:
...
+  if isinstance(value, type(tf.constant(0))):
+    value = value.numpy() # BytesList won't unpack a string from an EagerTensor.
+  return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
+
+def serialize_example(image: str) -> any:
+  """This function works to serialize an image path and returns a binary string
+  Args:
+    image: an image path to a jpg file
+  """
+  feature = {
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: This format doesn't match the format we use in `read_tfrecord`, so we will have to make format a parameter to `read_tfrecord`.

File: Code/monet_cyclegan/utils.py [hunk 2]
```diff
@@ -13,12 +13,13 @@
 import tensorflow as tf
 from PIL import Image
 from numpy import uint8, ndarray
+import cv2
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: Need to add `opencv-python` to Conda environment file.

File: Code/monet_cyclegan/data_acquisition/augment.py [hunk 3]
```diff
@@ -96,6 +96,20 @@ def augment_image(image: tf.Tensor) -> tf.Tensor:
...
+        image: The image to be augmented.
+
+    Returns:
+        The augmented image.
+    """
+    image = random_rotate(image)
+    image = random_flip(image)
+
+    return image
+
```
Comment (airicbear) [inline Code/monet_cyclegan/data_acquisition/augment.py]: We can just have options for the augment image function to include/exclude different augmentations such as crop.

File: Code/monet_cyclegan/data_acquisition/augment.py [hunk 4]
```diff
@@ -121,6 +135,30 @@ def save_augmented_image(input_path: str, output_dir: str) -> None:
...
+                       channels=CHANNELS)[0]
+
+    processed_image = augment_image_without_crop(image=image)
+    processed_image = tensor_to_image(image=processed_image)
+
+    filename = os.path.basename(input_path)
+
+    save_image(image=processed_image,
+               output_path=f'{output_dir}/augmented-{filename}')
+
```
Comment (airicbear) [inline Code/monet_cyclegan/data_acquisition/augment.py]: Same thing for augment_image_without_crop--we can just add parameters for the existing functions to include/exclude different augmentations.

File: Code/monet_cyclegan/utils.py [hunk 5]
```diff
@@ -284,4 +285,66 @@ def log_args(args: Namespace) -> None:
...
+        return "Error: The test file path does not exist"
+    number_of_values_in_train_directory = len(os.listdir(jpg_train_file_path))
+    ten_percent_of_values = float(number_of_values_in_train_directory)/10.0
+    ten_percent_rounded = round(ten_percent_of_values)
+    indexes_already_used = set()
+    while len(indexes_already_used) < ten_percent_rounded:
+        random_index_candidate = round(random_number(0,number_of_values_in_train_directory-1))
+        if random_index_candidate not in indexes_already_used:
+            picture_path_to_add_and_augment =jpg_train_file_path[random_index_candidate]
+            save_augmented_image_without_crop(jpg_train_file_path+picture_path_to_add_and_augment,jpg_test_file_path)
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: Circular import: `save_augmented_image_without_crop`

File: Code/monet_cyclegan/utils.py [hunk 6]
```diff
@@ -284,4 +285,66 @@ def log_args(args: Namespace) -> None:
     """
 
     for arg, value in sorted(vars(args).items()):
-        logger.info(f'{arg}: {value}')
+        logging.info(f'{arg}: {value}')
+
+def ten_percent_for_test(jpg_train_file_path: str, jpg_test_file_path:str) -> None:
+    """This function parses a file path to a train dataset that contains jpg files, alters 10% of them and then adds them to the test data set
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: How does it alter the 10%?

File: Code/monet_cyclegan/utils.py (hunk 1 above)
Comment (airicbear) [reply Code/monet_cyclegan/utils.py]: Or we can just only use `'image'` when reading TFRecord files.

File: Code/monet_cyclegan/utils.py [hunk 7]
```diff
@@ -284,4 +285,66 @@ def log_args(args: Namespace) -> None:
...
+    Args:
+        jpg_train_file_path: file path for the existing train file
+        jpg_test_file_path: file path for the existing test file
+    """
+    if os.path.exists(jpg_train_file_path)!=True:
+        return "Error: The train file path does not exist"
+    if os.path.exists(jpg_test_file_path)!=True:
+        return "Error: The test file path does not exist"
+    number_of_values_in_train_directory = len(os.listdir(jpg_train_file_path))
+    ten_percent_of_values = float(number_of_values_in_train_directory)/10.0
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: We can use percent parameter instead of hardcoding 10.0

File: Code/monet_cyclegan/utils.py [hunk 8]
```diff
@@ -284,4 +285,66 @@ def log_args(args: Namespace) -> None:
...
+def ten_percent_for_test(jpg_train_file_path: str, jpg_test_file_path:str) -> None:
+    """This function parses a file path to a train dataset that contains jpg files, alters 10% of them and then adds them to the test data set
+
+    Args:
+        jpg_train_file_path: file path for the existing train file
+        jpg_test_file_path: file path for the existing test file
+    """
+    if os.path.exists(jpg_train_file_path)!=True:
+        return "Error: The train file path does not exist"
+    if os.path.exists(jpg_test_file_path)!=True:
```
Comment (airicbear) [inline Code/monet_cyclegan/utils.py]: It doesn't make sense to ensure that the output directory exists. Instead we can prompt user to overwrite the files if the directory exists.

File: Code/monet_cyclegan/utils.py (hunk 7 above)
Comment (airicbear) [reply Code/monet_cyclegan/utils.py]: note: using 10 percent requires the dataset to be at least size 10.

File: Code/monet_cyclegan/utils.py (hunk 8 above)
Comment (airicbear) [reply Code/monet_cyclegan/utils.py]: We should prompt the user if they want to overwrite each existing file, not overwrite the entire directory.

File: Code/monet_cyclegan/utils.py (hunk 2 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: Yeah, I mentioned that in my pull request, I don't know how to add it to the Conda files.

File: Code/monet_cyclegan/utils.py (hunk 7 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: I used ten percent because that was the number she mentioned. This function will be applied to the training datasets, so they will always be of size 10.

File: Code/monet_cyclegan/utils.py (hunk 1 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: I think switching to using 'image' when reading TFRecord files will be easier. I believe we made this change in a Kaggle notebook that used these TFrecords https://www.kaggle.com/code/george1aeillo/monet-cyclegan-tutorial-489942-14ef5a?scriptVersionId=120655746

File: Code/monet_cyclegan/utils.py (hunk 6 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: It augments 10% of the image from the training file and then saves the augmented image to the test directory.

File: Code/monet_cyclegan/utils.py (hunk 8 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: The user isn't overwritting files, they are augmenting a file from the existing training data and then adds it to a new directory, using the augment function.

File: Code/monet_cyclegan/utils.py (hunk 5 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: I changed this by placing the ten_percent function in augment.py since it is related to augmenting the images.

File: Code/monet_cyclegan/data_acquisition/augment.py (hunk 3 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/data_acquisition/augment.py]: I added a crop parameter for this function in my latest commit.

File: Code/monet_cyclegan/data_acquisition/augment.py (hunk 4 above)
Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/data_acquisition/augment.py]: I added a crop parameter for this function in my latest commit.

File: Code/monet_cyclegan/utils.py (hunk 8 above)
Comment (airicbear) [reply Code/monet_cyclegan/utils.py]: Yeah I meant if we are running the function twice then there will be existing files. Also I figured we should be able to output the images even if the test set doesn't exist yet.

Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: Should we add a clean up part to the function that checks if any of the images in the test file have augmented in their names and if so deletes them. The save as augmented function I believe saves the files as augmented in the string name.

Comment (GeorgeFAeillo) [reply Code/monet_cyclegan/utils.py]: Why wouldn't the test set exist? Wouldn't it always exist when we use this function.

**Review by airicbear — APPROVED** [review]
